src/share/classes/com/sun/javatest/regtest/exec/ProcessCommand.java

```diff
@@ -243,15 +246,41 @@ public Status exec() {
             InputStream processErr = process.getErrorStream();
 
             long start = System.currentTimeMillis();
+            Alarm pre_alarm = Alarm.NONE;
             Alarm alarm = Alarm.NONE;
             final CountDownLatch timeoutHandlerDone = new CountDownLatch(1);
+            final long PRE_ALARM_TIME_MS = 5000;
 
             if (timeout > 0) {
                 final Thread victim = Thread.currentThread();
-                alarm = Alarm.schedule(timeout, TimeUnit.MILLISECONDS, out, new Runnable() {
-                    public void run() {
-                        invokeTimeoutHandler(timeoutHandler, timeoutHandlerDone, process, victim);
+                pre_alarm = Alarm.schedule(timeout, TimeUnit.MILLISECONDS, out, new Runnable() {
+                  // at the timeout, as the last resort, we will exit all the (possibly hanging),
+                  // descendants of the process, in the hope that doing so will unstuck the process itself
+                  public void run() {
+                    long pid = getTimeoutHandler().getProcessId(process);
+                    Optional<ProcessHandle> oph = ProcessHandle.of(pid);
+                    if (oph.isPresent()) {
+                      ProcessHandle processHandle = oph.get();
+                      Stream<ProcessHandle> descendants = processHandle.descendants();
+                      long count = descendants.count();
+                      if (count > 0) {
+                        System.err.println("Detected a possibly stuck process (pid:"+pid+") with "+count+" descendants");
```

Comment:
Don't use `System.err` directly;  use the `err` and `out` streams that were passed in when the class was invoked.

Reply:
Thank you, will fix.

Reply:
That got much more complicated that I imaged. Please take a look and tell me if I missed something and it could have been done in a simpler way.

Reply:
Hmm, I think I did overcomplicate things. After my last commit, I started working on the other feedback, and noticed "err" print writer, so I am using it now instead.